src/librustdoc/formats/cache.rs

```diff
@@ -127,9 +127,27 @@ crate struct Cache {
 /// This struct is used to wrap the `cache` and `tcx` in order to run `DocFolder`.
 struct CacheBuilder<'a, 'tcx> {
     cache: &'a mut Cache,
+    /// This field is used to prevent duplicated impl blocks.
+    impl_ids: FxHashMap<DefId, FxHashSet<DefId>>,
     tcx: TyCtxt<'tcx>,
 }
 
+/// This function is used to extract a `DefId` to be used as a key for the `Cache::impls` field.
+///
+/// It allows to prevent having duplicated implementations showing up (the biggest issue was with
+/// blanket impls).
+#[inline]
+fn expect_def_id(item_id: ItemId) -> DefId {
```

Comment:
This function is a landmine, because its type signature is a lie. You can't just go from `ItemId -> DefId`, you need to know what you're looking at.

* If we know that your `Impl::def_id` function will always return something that actually has a `DefId`, then this function shouldn't be necessary. Make that other function actually return a `DefId`.
* If you don't know that, have this function return an `Option` instead, and everywhere it's called, add a comment justifying why, in that particular spot, you know that primitives will never show up.
* If you really can't come up with a good justification, then it's fine. `ItemId` implements `Hash`, so you can just make it an `FxHashMap<DefId, FxHashSet<ItemId>>` instead.

Reply:
> * If we know that your `Impl::def_id` function will always return something that actually has a `DefId`, then this function shouldn't be necessary. Make that other function actually return a `DefId`.

This is a good point. That'll remove one level.

>     * If you really can't come up with a good justification, then it's fine. `ItemId` implements `Hash`, so you can just make it an `FxHashMap<DefId, FxHashSet<ItemId>>` instead.

That's what I went with originally and it didn't work because the `for` was always different, even though it was the same impl ID. It's the very root of this bug.